File: CodingSamples/DotNet/Fundamentals/Concurrency/ThreadingTest/DemoApp2/JointAccount.cs
namespace DemoApp;

class JointAccount
{
    public int Balance { get; private set; }

    public bool Debit(int amount)
    {
        bool success = false;
        Monitor.Enter(this); //lock monitor attached to 'this' object
        try
        {
            if(Balance >= amount)
            {
                Balance = Activity.Perform(Balance, amount, -1);
                success = true;
            }
        }
        finally
        {
            Monitor.Exit(this); //unlock monitor attached to'this' object
        }
        return success;
    }

    public void Credit(int amount)
    {
        lock(this)
        {
            Balance = Activity.Perform(Balance, amount, 1);
            //signal the monitor attached to this object
            Monitor.Pulse(this);
        }
    }

    public bool DebitAfterCredit(int amount)
    {
        lock(this)
        {
            //unlock monitor attached to this object and wait for
            //the monitor to be signalled by another thread
            Monitor.Wait(this);
            return Debit(amount);
        }
    }
}

File: CodingSamples/Java/Fundamentals/Concurrency/ThreadTest/demo2/app/JointAccount.java
package app;

public class JointAccount {
    
    private int balance;

    public int balance() {
        return balance;
    }

    public boolean debit(int amount) {
        boolean success = false;
        //to execute synchronized block a thread must lock the monitor
        //associated with the target (this) object, and this monitor
        //is unlocked at the end of the synchronized block
        synchronized(this){
            if(balance >= amount){
                balance = Activity.perform(balance, amount, -1);
                success = true;
            }
        }
        return success;
    }

    public synchronized void credit(int amount) {
        balance = Activity.perform(balance, amount, +1);
        this.notify(); //notify this object's monitor
    }

    public boolean debitAfterCredit(int amount) throws InterruptedException{
        synchronized(this){
            //wait on this object's monitor until it is notified
            //by some other thread
            this.wait();  
            return debit(amount);
        }
    }
}
